### ISSUE TITLE
Feature: inverse of `nom::combinator::cut`, possibly `uncut`

### ISSUE THREAD

**Issue by LoganDark** [Geal/nom#1527]
`nom` is currently missing an inverse of the `cut` combinator, making it unfortunately quite difficult to use `cut` without "infecting" the entire parser chain with cascading failure.

I ran into this issue while making my toy assembler, where inside integer parsing, I used `cut` to ensure that if the `0x` / `0b` tags were found, `alt` wouldn't go on to try decimal parsing if the hex/bin parsers fail (for obvious reasons, due to the prefix starting with `0`). The code was something like this:

```rust
	alt((
		preceded(tag(b"0x"), cut(hex_uint)),
		preceded(tag(b"0b"), cut(bin_uint)),
		dec_uint
	))(input)
```

However, using this as-is "infected" the parser chain, basically making the entire thing fail if this specific invocation found any `0x`/`0b`, and forbidding *any* ancestor `alt`s from continuing.

This does look like *exactly* the right use-case for `cut`, but the lack of an `uncut` made it work a bit *too* well. So I defined my own `uncut`:

```rust
pub fn uncut<I, O, E: ParseError<I>, F: Parser<I, O, E>>(mut parser: F) -> impl FnMut(I) -> IResult<I, O, E> {
	move |input: I| match parser.parse(input) {
		Err(Err::Failure(e)) => Err(Err::Error(e)),
		rest => rest,
	}
}
```

This allowed me to use `uncut` to localize `cut`'s effect, preventing the use of it from infecting the entire parser chain:

```rust
	uncut(alt((
		preceded(tag(b"0x"), cut(hex_uint)),
		preceded(tag(b"0b"), cut(bin_uint)),
		dec_uint
	)))(input)
```

This had the intended effect, and I would love to see (a version of) this combinator added to `nom` in the future. I'm willing to open a pull request if you like the idea.

###### I offer all code samples included in this comment under the MIT License.

**Comment by Stargateur** [comment]
This would catch every failure of any underline parser, it's probably a bad idea. I think you are trying to opposite to combinator concept with your `cut` usage.

Why not:

```rust
alt((
		dec_uint,
		preceded(tag(b"0x"), hex_uint),
		preceded(tag(b"0b"), bin_uint),
	))(input)
```

**Comment by Stargateur** [comment]
> that would immediately and incorrectly parse the 0 in 0xFF as a decimal 0 and then stop there, wouldn't it?

Yes, if `dec_uint` match one `0`, was just a random idea, just keep your alt and remove cut call, don't worry about error path, nom is mean to be used that way. If you really really want advanced control you can still do the alt manually using good old `if` and `return`.

> That's the point. I don't want a failure to bubble up and cause a top-level alt to stop trying alternative instructions.

Sorry, I wasn't clear, I mean this would catch failure of ALL underline parser, your use case is simple but what if `dec_uint` return a failure ? You would catch it, that probably not what you want.

**Comment by Stargateur** [comment]
Let's forget about your use case I think we mixed thing. That an interesting way to use Failure indeed. I have not yet see this pattern wanted to be used in nom but that could be a new thing.

**Comment by LoganDark** [comment]
> That an interesting way to use Failure indeed.

Is this not an idiomatic way to use Failure?

> Thus, I still think you don't see the corner case of this.

I know your so-called "corner case" which is `hex_uint`/`bin_uint` returning a `Failure` itself that is not from `cut`. But even in that case I would want to convert it to an `Error` anyway, because again, I don't need it to bubble all the way up to `uint`'s caller, I only want to use it to stop `uint`'s `alt` if it sees a tag and knows exactly which branch needs to be taken. This is even specified in `Failure`'s documentation: "we got to the right branch and we know other branches won't work". It's used that way in real `nom` code, it's just that not many people use `uncut` to localize it.

However, there are examples of others in the ecosystem implementing `uncut` independently (at least 3 on GitHub). Here are some:

- Localizing recognize_float failure, so that even if it fails early, it doesn't kill the entire parse chain

  https://github.com/HLTAS/hltas/blob/b44f60a4b55c7c0eaef48989c8d10843e1c28082/src/read/line.rs#L21-L31

- Another `uncut`:

  https://github.com/HenryHeffan/rust_act/blob/e67d19649ecd7db976143481218a9f501320ed6a/lex_parse/src/utils.rs#L893-L898

- Another `uncut`:

  https://github.com/kurtlawrence/kserd/blob/108321d103e1efd8f4851ca3bd94e4ce278fd423/src/parse/mod.rs#L116-L127

And this is just for those who named it `uncut`, I'm sure there are others who named it differently.

**Comment by Stargateur** [comment]
Thank it's a very good, argument, again you clearly make your point about uncut use case. But my point is that Failure is also used as runtime error of a parser like https://github.com/Geal/nom/blob/294ffb3d9e0ade2c3b7ddfff52484b6d643dcce1/src/multi/mod.rs#L734-L737

Meaning uncut would not only catch expected Failure but also not expected Failure, this could make you project hard to maintain.

The documentation of [Failure](https://docs.rs/nom/7.1.1/nom/enum.Err.html#variant.Failure )is currently:

> The parser had an unrecoverable error: we got to the right branch and we know other branches won’t work, so backtrack as fast as possible

Before your proposition I always see Failure as unrecoverable error, thus I think we may have a problem of clarity.

**Comment by LoganDark** [comment]
> Meaning uncut would not only catch expected Failure but also not expected Failure, this could make your project hard to maintain.

I thought I already clarified **at least twice** that I understand this perfectly well and I intentionally **want** to catch this "expected Failure". Here are some of my prior comments on the matter:

> `hex_uint`/`bin_uint` returning a `Failure` itself that is not from `cut`. But even in that case I would want to convert it to an `Error` anyway

> I do not want `dec_uint` to be able to abort the entire parser - **just the `uint` parser**

Is English not your first language? I'm curious about why you would have trouble understanding this.

`uncut` is used to catch failures, that's its point. Failures do their job with whatever is inside `uncut`, but there is a *reason* you would surround something in `uncut` rather than letting it bubble up. The reason does not change when there is an "expected Failure" because I still want to catch it and prevent it from aborting the entire parser.

> I always see Failure as unrecoverable error globally

There's no reason that Failure should be unrecoverable, it's more like an exception that should be able to be caught with `uncut`. For example in my case there is effectively an "exception" for "found 0x but invalid hex" and "found 0b but invalid binary" that prevents it from going on to decimal parsing, but I catch that in `uint` and transform it back into a normal error.

---

I'm not going to continue this discussion with you as you've just raised the same point repeatedly and ignored everything I said about it, I would much rather just wait for @Geal to weigh in on this and whether or not it would be a good fit for `nom`. Thank you and have a nice day, I apologize if I was rude at any point.

**Comment by Geal** [comment]
so, a bit of context on the `cut` combinator. The name is inspired from the [`cut`](http://www.cse.unsw.edu.au/~billw/dictionaries/prolog/cut.html) operator in prolog, that is used to prevent backtracking. So its usage in nom, and the `Failure` error variant were added for the case of preventing backtracking and alternative paths in `alt`. The meaning evolved to an unrecoverable error, like the way it is use in `many`.

But what makes an error unrecoverable depends on the context. In a web server, A TCP connection breaking is unrecoverable for the session, but a routine error for the server. Here in the parser, `Failure` will make that path of the parser unrecoverable, but ultimately the parser developer knows best how far that error can go, and can decide to recover in another part of the parser.

**So I believe the `uncut` combinator would be a nice addition (can't believe it was not added earlier), with some documentation about the tradeoff that it stops unrelated errors** (they would still be error though). I think there can be further discussion on naming (`cut` is not obvious) and semantics of those operators. Maybe that could be a kind of `try/catch`, where the `catch` part filters what kind of `Failure` it transforms?

**Comment by LoganDark** [comment]
> so, a bit of context on the `cut` combinator. The name is inspired from the [`cut`](http://www.cse.unsw.edu.au/~billw/dictionaries/prolog/cut.html) operator in prolog, that is used to prevent backtracking. So its usage in nom, and the `Failure` error variant were added for the case of preventing backtracking and alternative paths in `alt`. The meaning evolved to an unrecoverable error, like the way it is use in `many`.
> 
> But what makes an error unrecoverable depends on the context. In a web server, A TCP connection breaking is unrecoverable for the session, but a routine error for the server. Here in the parser, `Failure` will make that path of the parser unrecoverable, but ultimately the parser developer knows best how far that error can go, and can decide to recover in another part of the parser.
> 
> **So I believe the `uncut` combinator would be a nice addition (can't believe it was not added earlier), with some documentation about the tradeoff that it stops unrelated errors** (they would still be error though). I think there can be further discussion on naming (`cut` is not obvious) and semantics of those operators. Maybe that could be a kind of `try/catch`, where the `catch` part filters what kind of `Failure` it transforms?

Exactly! A fatal error in a sub-parser may be fatal for that specific parser, but an `uncut` is necessary in some cases to prevent the sub-parser from being a fatal error for upper parsers if it doesn't need to be.

There's nothing we can do about the current naming of `cut`, although you can definitely note something down for nom 8. In the meantime I'd love to get something merged into nom 7 to go out in the same release as #1526.

I'd love to chat with you about what the right name and usage would be for the new combinator. I'd like it to fit in with `cut` for the time being (so no `try`/`catch` - especially because of `try` being a reserved word) but it'll definitely be ready for a rework in nom 8. If the current `uncut` name and usage isn't optimal, what are your thoughts? We could pick something that fits in a little better with nom 7, for now.

**Comment by sshine** [comment]
I recently used `cut` to handle an unrecoverable error in a sub-parser nested inside a `many0`:

https://stackoverflow.com/questions/74993188/how-to-propagate-nom-fail-context-out-of-many0

In my case it was to disallow keywords inside labels, so the two scenarios are `<name>:` and `call <name>` with `<name>` being non-keyword. Because a keyword may also occur at the same level as `<name>:`, the keyword parser and label parser live in the same `alt` and share a common prefix, so if labels use `cut`, I have to try and parse keywords first.

I don't think a try-catch would help me here, since the problem was that `many0` hid the failing branch. [Megaparsec's `try`](https://hackage.haskell.org/package/megaparsec/docs/Text-Megaparsec.html#v:try) comes to mind. Since Megaparsec's `try` allows arbitrary lookahead, `try`ing a very big parser may be quite inefficient at a busy branch point, so it should be used with caution. Still, it does not infect the parser chain as @LoganDark nicely put it, it just introduces a performance overhead.

In my case, I went with using `cut`, but @Stargateur's suggestion to [restructure the parser](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=40c6fe511848d1e937b3100c307d70f9) so that I do my check *after* parsing is actually a lot nicer (it also allows for aggregating errors). As far as I remember, you can always avoid `cut` in Prolog with better design patterns, so the main reason to use it is performance (at the cost of infecting the search path).

**Comment by LoganDark** [comment]
I've had this page open for some weeks and I've been thinking about this for a while. It's been a long time and it still doesn't look like nom has this built-in yet. I've been tossing around the idea of calling it `scope` since the idea is to allow a sub-parser to fail without failing up the entire chain. I'm not entirely sold on calling it `catch` because we don't have a corresponding `try`, and also because `catch` might imply it catches errors as well? `scope` seems a bit better suited to the idea that the purpose isn't to handle errors but to contain failures.
